content/blog/2023-03-23-when-a-date-breaks-booting-the-kernel.md

```diff
@@ -0,0 +1,213 @@
+
+
+## But... why?
+
+OK, so it's obvious at this point that the timestamp is affecting loading a root disk somehow. But why? The obvious answer is that it's before the UNIX epoch. That something in the build process is turning the date into an actual timestamp, and going wrong when that timestamp gets used for something.
+
+But it's not like there was a build error complaining about it. As best I could tell, the kernel doesn't try to parse the date anywhere, besides passing it to `date` during the build. And if `date` had an issue with it, it would have brokem the _build_. Not _booting_ the kernel. There's no `date` utility being invoked during kernel boot!
+
+Regardless, I set about tracing the usage of `KBUILD_BUILD_TIMESTAMP` inside the kernel. The stacktrace in the panic gave the end point of the search; the function `mount_block_root()` wasn't happy. So all I had to do was work out at which point `mount_block_root()` tried to access the `KBUILD_BUILD_TIMESTAMP` value.
+
+In short, that went nowhere.
+
+`mount_block_root()` effectively just tries to open a file in the filesystem. There's massive amounts of code handling this, and any part could have had the undocumented dependency on `KBUILD_BUILD_TIMESTAMP`. Approaching from the other direction, `KBUILD_BUILD_TIMESTAMP` is turned into `build-timestamp` inside a Makefile, which is in turn related to a file `include/generated/utsversion.h`. This file `#define`s `UTS_VERSION` equal to the `KBUILD_BUILD_TIMESTAMP` value. Searching the kernel for `UTS_VERSION`, we hit `init/version-timestamp.c` which stores it in a struct with other build information:
+```C
+struct uts_namespace init_uts_ns = {
+	.ns.count = REFCOUNT_INIT(2),
+	.name = {
+		.sysname	= UTS_SYSNAME,
+		.nodename	= UTS_NODENAME,
+		.release	= UTS_RELEASE,
+		.version	= UTS_VERSION,
+		.machine	= UTS_MACHINE,
+		.domainname	= UTS_DOMAINNAME,
+	},
+	.user_ns = &init_user_ns,
+	.ns.inum = PROC_UTS_INIT_INO,
+#ifdef CONFIG_UTS_NS
+	.ns.ops = &utsns_operations,
+#endif
+};
+```
+
+This is where the trail goes cold: I don't know if you've ever tried this, but searching for `.version` in the kernel's codebase is not a very fruitful endeavor when you're interested in a specific kind of version.
+
+```text
+$ rg "(\.|\->)version\b" | wc -l
+5718
+```
+
+I tried tracing the usage of `init_uts_ns`, but didn't get very far.
+
+By now I'd already posted this in chat, and another developer Joel was also investigating this bizarre bug. They had been testing different timestamp values and made the horrifying discovery that the bug sticks around after a rebuild. So you could start with a broken build, set the timestamp back to the correct value, and the resulting kernel would _still be broken_. The boot log would report the correct time, but the root disk mounter panicked all the same.
```

Comment:
you can change "another developer Joel" to Joel with a link to either his sthbrx author page or his twitter account